Sources/SwiftFrameCore/Workers/ImageWriter.swift

```diff
@@ -0,0 +1,103 @@
+import AppKit
+import Foundation
+
+public final class ImageWriter {
+
+    // MARK: - Exporting
+
+    static func finish(
+        context: CGContext,
+        with outputPaths: [FileURL],
+        sliceSize: CGSize,
+        outputWholeImage: Bool,
+        locale: String,
+        suffix: String,
+        format: FileFormat,
+        completion: @escaping () throws -> Void) throws {
+        guard let image = context.makeImage() else {
+            throw NSError(description: "Could not render output image")
+        }
+        DispatchQueue.global().ky_asyncOrExit {
+            let slices = sliceImage(image, with: sliceSize)
+            var slicesFinished = false
+            var bigImageFinished = !outputWholeImage
+
+            let workGroup = DispatchGroup()
+            workGroup.enter()
+
+            // Writing images asynchronously gave a big performance boost, what a surprise
+            // Also, since we checked beforehand if the directory is writable, we can safely put of the rendering work to a different queue
+            DispatchQueue.global().ky_asyncOrExit {
+                try ImageWriter.write(images: slices, to: outputPaths, locale: locale, suffix: suffix, format: format)
+                slicesFinished = true
+                if slicesFinished && bigImageFinished {
+                    workGroup.leave()
+                }
+            }
+
+            if outputWholeImage {
+                DispatchQueue.global().ky_asyncOrExit {
+                    try outputPaths.forEach {
+                        try ImageWriter.write(image, to: $0.absoluteURL.appendingPathComponent(locale), fileName: "\(locale)-\(suffix)-big", format: format)
+                    }
+                    bigImageFinished = true
+                    if slicesFinished && bigImageFinished {
+                        workGroup.leave()
+                    }
+                }
+            }
+
+            _ = workGroup.wait(timeout: .now() + 5.00)
+            try completion()
+        }
+    }
+
+    static func sliceImage(_ image: CGImage, with size: CGSize) -> [CGImage] {
+        guard CGFloat(image.width).truncatingRemainder(dividingBy: size.width) == 0 else {
+            print("Image width is not a multiple in width of desired size")
+            return []
+        }
+        let numberOfSlices = image.width / Int(size.width)
+        var croppedImages = [CGImage?]()
+
+        for i in 0..<numberOfSlices {
+            let rect = CGRect(x: size.width * CGFloat(i), y: 0, width: size.width, height: size.height)
+            croppedImages.append(image.cropping(to: rect))
+        }
+        return croppedImages.compactMap { $0 }
```

Comment:
Should we catch `nil` elements here? That would just omit certain slices from the output which I think is an error case.